### PR TITLE
Selc 830 new onboarding fields

### PR DESCRIPTION
**Pull request by antonioT90**

### CODE REVIEW

File: src/main/resources/application-standalone.conf [hunk 1]
```diff
@@ -1,101 +1,113 @@
...
-    }
-  }
-
-  jwt {
-    public-keys {
-      url = ${WELL_KNOWN_URL}
-    }
-  }
-}
+include "kamon"
```
Comment (galales) [inline src/main/resources/application-standalone.conf]: probably the `new line` character has changed. Could you please check (in this file and in the others)?

File: src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala [hunk 2]
```diff
@@ -616,7 +687,9 @@ class ProcessApiServiceImpl(
         attributes = Seq(PartyManagementDependency.Attribute(category.origin, category.code, category.name)),
         products = Set.empty,
         address = institution.address,
-        zipCode = institution.zipCode
+        zipCode = institution.zipCode,
+        institutionType = "TBD",
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala]: if this information is not always available it is probably better to change the type to `Option[String]` and use a `None` here

File: src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala [hunk 3]
```diff
@@ -329,7 +329,17 @@ class ProcessApiServiceImpl(
...
+        createOrGetRelationship(
+          person.id,
+          institution.id,
+          roleToDependency(role),
+          product,
+          productRole,
+          if (isManager) onboardingRequest.pricingPlan else Option.empty,
+          if (isManager) onboardingRequest.institutionUpdate else Option.empty,
+          if (isManager) onboardingRequest.billing else Option.empty
+        )(bearer)
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala]:
```suggestion
       role match {
         case PartyRole.MANAGER =>
           createOrGetRelationship(
          person.id,
          institution.id,
          roleToDependency(role),
          product,
          productRole,
          onboardingRequest.pricingPlan,
          onboardingRequest.institutionUpdate,
          onboardingRequest.billing
        )(bearer)
        case _ => 
          createOrGetRelationship(
          person.id,
          institution.id,
          roleToDependency(role),
          product,
          productRole,
          None,
          None,
          None
        )(bearer)
       }
```
sorry for the ugly formatting 😅

File: src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala [hunk 4]
```diff
@@ -367,7 +377,48 @@ class ProcessApiServiceImpl(
       ApplicationConfiguration.onboardingMailTaxCodePlaceholder     -> currentUser.externalId
     )
 
-    val bodyParameters: Map[String, String] = tokenParameters ++ userParameters
+    val institutionInfoParameters: scala.collection.mutable.Map[String, String] = scala.collection.mutable.Map()
+    if (onboardingRequest.institutionUpdate != null) {
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala]: you don't need this check, `map` executes the given function only if the field is not `None`.

(btw in Scala an empty `Option` is `None`, not `null`)

File: src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala [hunk 5]
```diff
@@ -367,7 +377,48 @@ class ProcessApiServiceImpl(
...
+      })
+    }
+
+    val billingParameters: scala.collection.mutable.Map[String, String] = scala.collection.mutable.Map()
+    if (onboardingRequest.pricingPlan != null) {
+      onboardingRequest.pricingPlan.map(o =>
+        billingParameters.put(ApplicationConfiguration.onboardingMailBillingPricingPlanPlaceholder, o)
+      )
+    }
+    if (onboardingRequest.billing != null) {
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala]: Same considerations for `null` check

File: src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala [hunk 6]
```diff
@@ -367,7 +377,48 @@ class ProcessApiServiceImpl(
...
+        )
+      })
+    }
+
+    val billingParameters: scala.collection.mutable.Map[String, String] = scala.collection.mutable.Map()
+    if (onboardingRequest.pricingPlan != null) {
+      onboardingRequest.pricingPlan.map(o =>
+        billingParameters.put(ApplicationConfiguration.onboardingMailBillingPricingPlanPlaceholder, o)
+      )
+    }
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala]: in Scala mutability is discouraged and should be avoided every time it is possible.

```suggestion
    val billingParameters: Map[String, String] = 
      onboardingRequest.pricingPlan.fold(Map.empty[String, String])(o =>
        Map(ApplicationConfiguration.onboardingMailBillingPricingPlanPlaceholder -> o)
      )

```

Same goes for `institutionInfoParameters`

**Review by galales — CHANGES_REQUESTED** [review]

File: src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala [hunk 7]
```diff
@@ -367,7 +377,48 @@ class ProcessApiServiceImpl(
       ApplicationConfiguration.onboardingMailTaxCodePlaceholder     -> currentUser.externalId
     )
 
-    val bodyParameters: Map[String, String] = tokenParameters ++ userParameters
+    val institutionInfoParameters: scala.collection.mutable.Map[String, String] = scala.collection.mutable.Map()
```
Comment (beetlecrunch) [inline src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala:380]:
```suggestion
    val institutionInfoParameters: Map[String, String] = {
      onboardingRequest.institutionUpdate.fold(Map.empty[String, String]) { iu =>
        Seq(
          iu.institutionType.zip(ApplicationConfiguration.onboardingMailInstitutionInfoInstitutionTypePlaceholder.some),
          iu.description.zip(ApplicationConfiguration.onboardingMailInstitutionInfoDescriptionPlaceholder.some),
          iu.digitalAddress.zip(ApplicationConfiguration.onboardingMailInstitutionInfoDigitalAddressPlaceholder.some),
          iu.address.zip(ApplicationConfiguration.onboardingMailInstitutionInfoAddressPlaceholder.some),
          iu.taxCode.zip(ApplicationConfiguration.onboardingMailInstitutionInfoTaxCodePlaceholder.some)
        ).flatten.toMap
      }
    }
```

Comment (beetlecrunch) [inline src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala]:
```suggestion
    val institutionInfoParameters: Map[String, String] = {
      onboardingRequest.institutionUpdate.fold(Map.empty[String, String]) { iu =>
        Seq(
          ApplicationConfiguration.onboardingMailInstitutionInfoInstitutionTypePlaceholder.some.zip(iu.institutionType),
          ApplicationConfiguration.onboardingMailInstitutionInfoDescriptionPlaceholder.some.zip(iu.description),
          ApplicationConfiguration.onboardingMailInstitutionInfoDigitalAddressPlaceholder.some.zip(iu.digitalAddress),
          ApplicationConfiguration.onboardingMailInstitutionInfoAddressPlaceholder.some.zip(iu.address),
          ApplicationConfiguration.onboardingMailInstitutionInfoTaxCodePlaceholder.some.zip(iu.taxCode)
        ).flatten.toMap
      }
    }
```

File: src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala (hunk 2 above)
Comment (antonioT90) [reply src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala]: This info will be filled after the first complete onboarding of the first product, only before this moment this field will not contains a value.
Using a dummy value will allow us to declare it as required, in order to not perform checks to verify if null when asking for data

File: src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala (hunk 7 above)
Comment (antonioT90) [reply src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala]: applied

File: src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala (hunk 6 above)
Comment (antonioT90) [reply src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala]: applied

File: src/main/resources/application-standalone.conf (hunk 1 above)
Comment (antonioT90) [reply src/main/resources/application-standalone.conf]: resolved

File: src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala (hunk 2 above)
Comment (antonioT90) [reply src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala]: @galales could I check as resolved this conversation?

Comment (galales) [reply src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala]: I see your point to have this field as required, but I'm still concerned about placeholders that are not enforced by the type system, and that need to be carefully managed.

That said, if this is your requirement and you have verified that all required validations and tests are in place, from my side you can resolve the conversation

Comment (antonioT90) [reply src/main/scala/it/pagopa/interop/partyprocess/api/impl/ProcessApiServiceImpl.scala]: As the release date has been postponed, we decided to let this field to be optional, as requested

**Pull request merged**